File: app/src/utils/daily-values.ts
import type { NutrientKey } from '../types';
import { NUTRIENT_META } from './nutrition-meta';

export interface UserProfile {
  sex: 'male' | 'female';
  weight_kg: number;
  height_cm: number;
  age: number | null;
}

type AgeRange = '19-30' | '31-50' | '51+';

function getAgeRange(age: number | null): AgeRange {
  if (age === null || age <= 30) return '19-30';
  if (age <= 50) return '31-50';
  return '51+';
}

const SEX_AGE_RDA: Record<
  'male' | 'female',
  Record<AgeRange, Partial<Record<NutrientKey, number>>>
> = {
  male: {
    '19-30': {
      iron_mg: 8,
      magnesium_mg: 400,
      zinc_mg: 11,
      potassium_mg: 3400,
      fiber_g: 38,
      vitamin_a_mcg: 900,
      vitamin_c_mg: 90,
      vitamin_k_mcg: 120,
      manganese_mg: 2.3,
      vitamin_b1_mg: 1.2,
      vitamin_b2_mg: 1.3,
      vitamin_b3_mg: 16,
      vitamin_b6_mg: 1.3,
    },
    '31-50': {
      iron_mg: 8,
      magnesium_mg: 420,
      zinc_mg: 11,
      potassium_mg: 3400,
      fiber_g: 38,
      vitamin_a_mcg: 900,
      vitamin_c_mg: 90,
      vitamin_k_mcg: 120,
      manganese_mg: 2.3,
      vitamin_b1_mg: 1.2,
      vitamin_b2_mg: 1.3,
      vitamin_b3_mg: 16,
      vitamin_b6_mg: 1.3,
    },
    '51+': {
      iron_mg: 8,
      magnesium_mg: 420,
      zinc_mg: 11,
      potassium_mg: 3400,
      fiber_g: 30,
      vitamin_a_mcg: 900,
      vitamin_c_mg: 90,
      vitamin_k_mcg: 120,
      manganese_mg: 2.3,
      vitamin_b1_mg: 1.2,
      vitamin_b2_mg: 1.3,
      vitamin_b3_mg: 16,
      vitamin_b6_mg: 1.7,
    },
  },
  female: {
    '19-30': {
      iron_mg: 18,
      magnesium_mg: 310,
      zinc_mg: 8,
      potassium_mg: 2600,
      fiber_g: 25,
      vitamin_a_mcg: 700,
      vitamin_c_mg: 75,
      vitamin_k_mcg: 90,
      manganese_mg: 1.8,
      vitamin_b1_mg: 1.1,
      vitamin_b2_mg: 1.1,
      vitamin_b3_mg: 14,
      vitamin_b6_mg: 1.3,
    },
    '31-50': {
      iron_mg: 18,
      magnesium_mg: 320,
      zinc_mg: 8,
      potassium_mg: 2600,
      fiber_g: 25,
      vitamin_a_mcg: 700,
      vitamin_c_mg: 75,
      vitamin_k_mcg: 90,
      manganese_mg: 1.8,
      vitamin_b1_mg: 1.1,
      vitamin_b2_mg: 1.1,
      vitamin_b3_mg: 14,
      vitamin_b6_mg: 1.3,
    },
    '51+': {
      iron_mg: 8,
      magnesium_mg: 320,
      zinc_mg: 8,
      potassium_mg: 2600,
      fiber_g: 21,
      vitamin_a_mcg: 700,
      vitamin_c_mg: 75,
      vitamin_k_mcg: 90,
      manganese_mg: 1.8,
      vitamin_b1_mg: 1.1,
      vitamin_b2_mg: 1.1,
      vitamin_b3_mg: 14,
      vitamin_b6_mg: 1.5,
    },
  },
};

const ACTIVITY_MULTIPLIER = 1.55;

export function computeProfileDailyValues(
  profile: UserProfile
): Partial<Record<NutrientKey, number>> {
  const result: Partial<Record<NutrientKey, number>> = {};
  const ageRange = getAgeRange(profile.age);
  const rdaTable = SEX_AGE_RDA[profile.sex][ageRange];

  for (const [key, value] of Object.entries(rdaTable)) {
    result[key as NutrientKey] = value;
  }

  result.protein_g = Math.round(profile.weight_kg * 0.8);

  if (profile.age !== null) {
    let bmr: number;
    if (profile.sex === 'male') {
      bmr = 10 * profile.weight_kg + 6.25 * profile.height_cm - 5 * profile.age + 5;
    } else {
      bmr = 10 * profile.weight_kg + 6.25 * profile.height_cm - 5 * profile.age - 161;
    }
    result.calories_kcal = Math.round(bmr * ACTIVITY_MULTIPLIER);
  }

  return result;
}

export type EffectiveDailyValues = Map<NutrientKey, number | null>;

export function getEffectiveDailyValues(
  profile: UserProfile | null,
  overrides: Partial<Record<NutrientKey, number>>
): EffectiveDailyValues {
  const profileValues = profile ? computeProfileDailyValues(profile) : {};
  const result = new Map<NutrientKey, number | null>();

  for (const meta of NUTRIENT_META) {
    const override = overrides[meta.key];
    if (override !== undefined) {
      result.set(meta.key, override);
      continue;
    }

    const profileValue = profileValues[meta.key];
    if (profileValue !== undefined) {
      result.set(meta.key, profileValue);
      continue;
    }

    result.set(meta.key, meta.dailyValue);
  }

  return result;
}

File: ios/Nutritionmaxx/Nutritionmaxx/Services/DailyValues.swift
import Foundation

struct UserProfile: Codable {
    var sex: BiologicalSex
    var weightKg: Double
    var heightCm: Double
    var age: Int?

    enum BiologicalSex: String, CaseIterable, Codable, Identifiable {
        case male, female
        var id: String { rawValue }
        var label: String {
            switch self {
            case .male: "Male"
            case .female: "Female"
            }
        }
    }
}

private enum AgeRange {
    case young, middle, senior
}

private func getAgeRange(_ age: Int?) -> AgeRange {
    guard let age else { return .young }
    if age <= 30 { return .young }
    if age <= 50 { return .middle }
    return .senior
}

private let sexAgeRDA: [UserProfile.BiologicalSex: [AgeRange: [NutrientKey: Double]]] = [
    .male: [
        .young: [
            .ironMg: 8, .magnesiumMg: 400, .zincMg: 11, .potassiumMg: 3400, .fiberG: 38,
            .vitaminAMcg: 900, .vitaminCMg: 90, .vitaminKMcg: 120, .manganeseMg: 2.3,
            .vitaminB1Mg: 1.2, .vitaminB2Mg: 1.3, .vitaminB3Mg: 16, .vitaminB6Mg: 1.3,
        ],
        .middle: [
            .ironMg: 8, .magnesiumMg: 420, .zincMg: 11, .potassiumMg: 3400, .fiberG: 38,
            .vitaminAMcg: 900, .vitaminCMg: 90, .vitaminKMcg: 120, .manganeseMg: 2.3,
            .vitaminB1Mg: 1.2, .vitaminB2Mg: 1.3, .vitaminB3Mg: 16, .vitaminB6Mg: 1.3,
        ],
        .senior: [
            .ironMg: 8, .magnesiumMg: 420, .zincMg: 11, .potassiumMg: 3400, .fiberG: 30,
            .vitaminAMcg: 900, .vitaminCMg: 90, .vitaminKMcg: 120, .manganeseMg: 2.3,
            .vitaminB1Mg: 1.2, .vitaminB2Mg: 1.3, .vitaminB3Mg: 16, .vitaminB6Mg: 1.7,
        ],
    ],
    .female: [
        .young: [
            .ironMg: 18, .magnesiumMg: 310, .zincMg: 8, .potassiumMg: 2600, .fiberG: 25,
            .vitaminAMcg: 700, .vitaminCMg: 75, .vitaminKMcg: 90, .manganeseMg: 1.8,
            .vitaminB1Mg: 1.1, .vitaminB2Mg: 1.1, .vitaminB3Mg: 14, .vitaminB6Mg: 1.3,
        ],
        .middle: [
            .ironMg: 18, .magnesiumMg: 320, .zincMg: 8, .potassiumMg: 2600, .fiberG: 25,
            .vitaminAMcg: 700, .vitaminCMg: 75, .vitaminKMcg: 90, .manganeseMg: 1.8,
            .vitaminB1Mg: 1.1, .vitaminB2Mg: 1.1, .vitaminB3Mg: 14, .vitaminB6Mg: 1.3,
        ],
        .senior: [
            .ironMg: 8, .magnesiumMg: 320, .zincMg: 8, .potassiumMg: 2600, .fiberG: 21,
            .vitaminAMcg: 700, .vitaminCMg: 75, .vitaminKMcg: 90, .manganeseMg: 1.8,
            .vitaminB1Mg: 1.1, .vitaminB2Mg: 1.1, .vitaminB3Mg: 14, .vitaminB6Mg: 1.5,
        ],
    ],
]

func computeProfileDailyValues(_ profile: UserProfile) -> [NutrientKey: Double] {
    var result: [NutrientKey: Double] = [:]
    let ageRange = getAgeRange(profile.age)

    if let rdaTable = sexAgeRDA[profile.sex]?[ageRange] {
        for (key, value) in rdaTable {
            result[key] = value
        }
    }

    result[.proteinG] = (profile.weightKg * 0.8).rounded()

    if let age = profile.age {
        let bmr: Double
        if profile.sex == .male {
            bmr = 10 * profile.weightKg + 6.25 * profile.heightCm - 5 * Double(age) + 5
        } else {
            bmr = 10 * profile.weightKg + 6.25 * profile.heightCm - 5 * Double(age) - 161
        }
        result[.caloriesKcal] = (bmr * 1.55).rounded()
    }

    return result
}

typealias EffectiveDailyValues = [NutrientKey: Double?]

func getEffectiveDailyValues(
    profile: UserProfile?,
    overrides: [NutrientKey: Double]
) -> EffectiveDailyValues {
    let profileValues = profile.map { computeProfileDailyValues($0) } ?? [:]
    var result: EffectiveDailyValues = [:]

    for meta in nutrientMetaList {
        if let override = overrides[meta.key] {
            result[meta.key] = override
            continue
        }
        if let profileValue = profileValues[meta.key] {
            result[meta.key] = profileValue
            continue
        }
        result[meta.key] = meta.dailyValue
    }

    return result
}

func resolveDV(_ meta: NutrientMeta, dvMap: EffectiveDailyValues?) -> Double? {
    if let dvMap, let val = dvMap[meta.key] {
        return val
    }
    return meta.dailyValue
}
